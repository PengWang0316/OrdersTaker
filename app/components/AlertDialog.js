import React from 'react';
import PropTypes from 'prop-types';
import { Dialog, DialogTitle, DialogContent, DialogContentText, DialogActions, Button } from '@material-ui/core';

export const AlertDialog = ({
  open, onClose, title, content, onSecondButton, onFirstButton, firstButtonText, secondButtonText
}) => (
  <Dialog
    open={open}
    onClose={onClose}
    aria-labelledby="alert-dialog-title"
    aria-describedby="alert-dialog-description"
  >
    <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
    <DialogContent>
      <DialogContentText id="alert-dialog-description">{content}</DialogContentText>
    </DialogContent>
    <DialogActions>
      <Button onClick={onFirstButton} color="primary">
        {firstButtonText}
      </Button>
      <Button onClick={onSecondButton} color="primary" autoFocus>
        {secondButtonText}
      </Button>
    </DialogActions>
  </Dialog>
);
AlertDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  onFirstButton: PropTypes.func.isRequired,
  onSecondButton: PropTypes.func.isRequired,
  firstButtonText: PropTypes.string,
  secondButtonText: PropTypes.string
};
AlertDialog.defaultProps = {
  firstButtonText: 'Cancel',
  secondButtonText: 'Confirm'
};
export default AlertDialog;
