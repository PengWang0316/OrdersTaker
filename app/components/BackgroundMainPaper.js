import React from 'react';
import PropTypes from 'prop-types';
import { withStyles, Paper } from '@material-ui/core';

const styles = {
  root: {
    backgroundImage: 'linear-gradient(to right top, #051937, #004d7a, #008793, #00bf72, #a8eb12)',
    width: '100%',
    height: '100%',
    padding: '0 20px'
  },
  paper: {
    margin: '40px auto 20px auto',
    maxWidth: 1024,
    borderRadius: '10px'
  }
};

export const BackgroundMainPaper = ({ classes, children }) => (
  <div className={classes.root}>
    <Paper className={classes.paper} elevation={4}>
      {children}
    </Paper>
  </div>
);
BackgroundMainPaper.propTypes = {
  classes: PropTypes.object.isRequired,
  children: PropTypes.element.isRequired
};
export default withStyles(styles)(BackgroundMainPaper);
