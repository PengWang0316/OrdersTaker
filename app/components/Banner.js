import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Hidden from '@material-ui/core/Hidden';
import { withStyles } from '@material-ui/core/styles';

import { fetchBasicInformation } from '../actions/BasicInformationActions';

const styles = {
  divPlaceHolder: {
    background: '#efefef',
    width: '100%',
    minHeight: '400px'
  },
  bigBanner: {
    width: '100%',
    maxHeight: '770px'
  },
  smallBanner: {
    width: '100%',
    maxHeight: '400px'
  },
  bannerText: {
    display: 'flex',
    flexDirection: 'column',
    width: '60%',
    position: 'absolute',
    marginTop: '18%',
    left: '20%'
  },
  title: {
    fontSize: '40px',
    color: 'white',
    textAlign: 'center',
    textTransform: 'uppercase',
    textShadow: '1px 1px 1px #909090'
  },
  subTitle: {
    color: 'white',
    fontSize: '24px',
    textAlign: 'center',
    textShadow: '1px 1px 1px #909090'
  },
  '@media (max-width: 600px)': {
    bannerText: { marginTop: '21%' },
    title: { fontSize: '30px' },
    subTitle: { fontSize: '18px' },
    divPlaceHolder: { minHeight: '230px' }
  }
};

/**
 * The Component to show the banner image and title.
 */
export class Banner extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    basicInformation: PropTypes.object,
    fetchBasicInformation: PropTypes.func.isRequired
  };
  static defaultProps = { basicInformation: null };

  /**
   * Call the redux action to get the basic infomariont.
   * @param {object} props contains this component's prop values;
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (!props.basicInformation) props.fetchBasicInformation();
  }

  /**
   * Render the jsx for the component.
   * @return {jsx} Return the jsx.
   */
  render() {
    const { classes, basicInformation } = this.props;
    return (
      <div className={basicInformation ? '' : classes.divPlaceHolder}>
        {basicInformation &&
          <Fragment>
            <div className={classes.bannerText}>
              <div className={classes.title}>{basicInformation.banners[0].title}</div>
              <div className={classes.subTitle}>{basicInformation.banners[0].subTitle}</div>
            </div>
            <Hidden only="xs">
              <img alt="Big Banner" src={basicInformation.banners[0].url} className={classes.bigBanner} />
            </Hidden>
            <Hidden only={['lg', 'md', 'sm']}>
              <img alt="Small Banner" src={basicInformation.banners[0].xsUrl} className={classes.smallBanner} />
            </Hidden>
          </Fragment>}
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = ({ basicInformation }) => ({ basicInformation });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchBasicInformation: () => dispatch(fetchBasicInformation())
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Banner));
