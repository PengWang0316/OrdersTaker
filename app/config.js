export const JWT_MESSAGE = 'OrdersTaker_JWT';
export const ADMINISTRATOR_ROLE = 1;
export const SUPER_USER_ROLE = 2;
export const NORMAL_USER_ROLE = 3;
// export const BASE_URL = 'https://orders-taker.glitch.me'; // Development server
// export const BASE_URL = 'https://orderstaker.kevin-project.com:8081'; // Development server on the AWS
export const BASE_URL = 'https://orderstaker.kevin-project.com:8080'; // Production server
// export const SOCKETIO_URL = 'https://orderstaker.kevin-project.com:3001';
export const SOCKETIO_URL = 'https://orderstaker.kevin-project.com:3000'; // Production server

export const CLOUDINARY_UPLOAD_PRESET = 'OrdersTakerWebApp'; // The preset of Cloudinary.
export const LOGIN_CALLBACK_URL = 'OrdersTaker_loginCallbackUrl';

export const ITEM_ONE_PRICE_KEY = '_onePrice';
export const MAX_ORDER_AMOUNT = 6; // How many orders will show in the one page
export const ORDER_STATUS_FINISHED = 'Finished';
export const ORDER_STATUS_RECEIVED = 'Received';
export const ORDER_STATUS_PREPARING = 'Preparing';

/* Lazy image loading */
export const LAZY_IMAGE_CLASS = 'lazyImage';
export const IMAGE_PLACEHOLDER_URL = 'https://res.cloudinary.com/orderstaker/image/upload/v1536616770/others/PlaceHolder.png';

/* URLS */
export const HOME_PAGE_URL = '/';
export const LOGIN_REDIRECT_RUL = '/loginRedirect';
export const CART_PAGE_URL = '/cartPage';
export const ORDERS_PAGE_URL = '/ordersPage';
export const KITHEN_INTERFACE_PAGE_URL = '/kithenInterfacePage';

/* Socket IO events */
export const SOCKETIO_EVENT_ADD_NEW_ORDER = 'socketioEventAddNewOrder';
export const SOCKETIO_EVENT_UPDATE_ORDER_ITEM = 'socketioEventUpdateOrderItem';
