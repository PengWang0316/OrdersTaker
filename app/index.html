<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="A simply order taking system for restaurants">
    <meta name="keywords" content="Order taker restaurant free">
    <!-- Add manifest -->
    <link rel="manifest" href="/manifest.json">
    <!-- Tell the browser it's a PWA -->
    <meta name="mobile-web-app-capable" content="yes">
    <!-- Tell iOS it's a PWA -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="theme-color" content="#6d9cb5"/>
    <link rel="icon" href="/favicon.png" >
    <!-- Latest compiled and minified CSS -->

    <title>Orders Taker - A simply order taking system for restaurants</title>
  </head>
  <body>
    <div id="app"></div>
    <noscript>This site requires JavaScript to work. Please change to a browser that supports JavaScript or turn on JavaScript in the setting.</noscript>
    <!-- Latest compiled and minified JavaScript -->
  </body>
</html>
