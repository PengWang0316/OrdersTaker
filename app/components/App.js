
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
// import Loadable from 'react-loadable'; Does not work with react-hot-loader very well.
// import universal from 'react-universal-component'; Does not work with react-hot-loader very well.
import importedComponent from 'react-imported-component';
import CssBaseline from '@material-ui/core/CssBaseline';

import LoginDialog from './LoginDialog/';
import LoginDialogSnackbar from './snackbars/LoginDialogSnackbar';
import LogoutSnackbar from './snackbars/LogoutSnackbar';
import Navbar from './Navbar';
import { parserUserFromJwt } from '../actions/UserActions';
import LoginDialogContext from '../contexts/LoginDialogContext';

import { JWT_MESSAGE, HOME_PAGE_URL, LOGIN_REDIRECT_RUL, ORDER_PAGE_URL, ORDER_STATUS_PAGE_URL, ORDERS_PAGE_URL } from '../config';
import LoadingAnimation from './SharedComponents/LoadingAnimation';

// import { CSSTransitionGroup } from 'react-transition-group';

// import BackgroundMainPager from './BackgroundMainPager';

const theme = createMuiTheme({
  palette: {
    primary: {
      light: '#9ecde7',
      main: '#6d9cb5',
      dark: '#3e6e85',
      contrastText: '#fff',
    },
    secondary: {
      light: '#497ca4',
      main: '#105075',
      dark: '#002884',
      contrastText: '#fff',
    },
  },
});

/* istanbul ignore next */
const HomePage = importedComponent(() => import(/* webpackChunkName: "HomePageContainer" */ './containers/HomePageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const LoginRedirect = importedComponent(() => import(/* webpackChunkName: "LoginRedirect" */ './LoginRedirect').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderPage = importedComponent(() => import(/* webpackChunkName: "OrderPageContainer" */ './containers/OrderPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderStatusPage = importedComponent(() => import(/* webpackChunkName: "OrderStatusPageContainer" */ './containers/OrderStatusPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrdersPage = importedComponent(() => import(/* webpackChunkName: "OrdersPageContainer" */ './containers/OrdersPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/**
 * The root component that contains the theme, routers, navbar, and login dialog
 */
export class App extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    parserUserFromJwt: PropTypes.func.isRequired
  };

  /**
   * If Redux does not have user and local storage has jwt, use jwt to parser the user.
   * @param {object} props contains the component's prop value.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (!props.user._id) {
      const jwtMessage = localStorage.getItem(JWT_MESSAGE);
      if (jwtMessage) props.parserUserFromJwt(jwtMessage);
    }
  }

  state = {
    isLoginDialogOpen: false,
    isLogoutSnackBarOpen: false,
    isLoginSnackbarOpen: false,
    snackbarMessage: ''
  };

  /**
   * Setting the open state to an opposite value.
   * @return {null} No return.
   */
  handleToggleDialog = () => this.setState(({ isLoginDialogOpen }) => ({ isLoginDialogOpen: !isLoginDialogOpen }));

  /**
   * Setting the snackbarOpen state to true in order to open the snackbar.
   * If message paramter is a string, set it to message state. Otherwise, set a empty message.
   * @param {string} message will be showed in the snackbar.
   * @return {null} No return.
   */
  handleToggleSnackbar = message => this.setState(({ isLoginSnackbarOpen }) =>
    ({ isLoginSnackbarOpen: !isLoginSnackbarOpen, snackbarMessage: typeof message === 'string' ? message : '' }));

  /**
   * Set the logoutSnackbarOpen state to an opposite value.
   * @return {null} No return.
   */
  handleToggleLogoutSnackbar = () => this.setState(({ isLogoutSnackBarOpen }) =>
    ({ isLogoutSnackBarOpen: !isLogoutSnackBarOpen }));

  /**
   * It will be pass down in a context in order to make it available for all component.
   * This method will be used when logout action is performed.
   * Clear the state isLoginDialogOpen in order to prevent the LoginDialog shows up.
   * @return {null} No return.
   */
  handleLogoutAction = () => this.setState({ isLoginDialogOpen: false, isLogoutSnackBarOpen: true });

  /**
   * The render method for the component
   * @return {jsx} Return jsx
   */
  render() {
    const { user } = this.props;
    const { isLogoutSnackBarOpen, isLoginDialogOpen, isLoginSnackbarOpen, snackbarMessage } = this.state;
    return (
      <MuiThemeProvider theme={theme}>
        <Router>
          {/* Using a context to give children components access to login dialog */}
          <LoginDialogContext.Provider value={{ handleLogoutAction: this.handleLogoutAction, handleToggleLoginDialog: this.handleToggleDialog }}>
            <div>
              <CssBaseline />
              <Navbar />
              <main>
                <Switch>
                  <Route exact path={HOME_PAGE_URL} component={HomePage} />
                  <Route exact path={ORDER_PAGE_URL} component={OrderPage} />
                  <Route exact path={LOGIN_REDIRECT_RUL} component={LoginRedirect} />
                  <Route path={`${ORDER_STATUS_PAGE_URL}/:orderId`} component={OrderStatusPage} />
                  <Route path={ORDERS_PAGE_URL} component={OrdersPage} />
                  <Route render={() => <p>Not Fount!</p>} />
                </Switch>
              </main>
              {!user._id && ( // If a user has already login, do not need to load componnets below.
                <Fragment>
                  <LoginDialog open={isLoginDialogOpen} onClose={this.handleToggleDialog} onToggleSnackbar={this.handleToggleSnackbar} />
                  <LoginDialogSnackbar open={isLoginSnackbarOpen} onClose={this.handleToggleSnackbar} message={snackbarMessage} />
                </Fragment>
              )}
              <LogoutSnackbar open={isLogoutSnackBarOpen} onClose={this.handleToggleLogoutSnackbar} />
            </div>
          </LoginDialogContext.Provider>
        </Router>
      </MuiThemeProvider>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  parserUserFromJwt: jwtMessage => dispatch(parserUserFromJwt(jwtMessage))
});
export default connect(mapStateToProps, mapDispatchToProps)(App);
