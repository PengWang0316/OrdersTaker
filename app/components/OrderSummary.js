import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';
import { red } from '@material-ui/core/colors';
import { RemoveShoppingCart, Send } from '@material-ui/icons';
import {
  Card, CardContent, Button, Typography, List, CircularProgress,
  ListItem, ListItemText, ListItemSecondaryAction, Tooltip
} from '@material-ui/core';

import QRCodeScanner from './QRCodeScanner/';
import { clearOrders, placeOrder } from '../actions/OrdersActions';
import AlertDialog from './AlertDialog';
import { HOME_PAGE_URL, ORDER_STATUS_PAGE_URL } from '../config';

/* istanbul ignore next */
const styles = theme => ({
  card: {
    padding: '10px 5px'
  },
  title: {
    fontWeight: 'bold',
    maringBottom: 20
  },
  categoryTitle: {
    fontSize: 14
  },
  flexEndDiv: {
    display: 'flex',
    justifyContent: 'flex-end'
  },
  fontBold: {
    fontWeight: 'bold',
    textDecoration: 'underline'
  },
  hrStyle: {
    border: 0,
    borderBottom: '1px dotted #ccc',
    background: '#999',
    margin: '10px 0'
  },
  placeBtn: {
    fontSize: 12
  },
  clearBtn: {
    fontSize: 12,
    margin: '0 10px',
    color: theme.palette.getContrastText(red[700]),
    backgroundColor: red[700],
    '&:hover': {
      backgroundColor: red[900]
    }
  },
  icon: {
    marginLeft: 6,
    fontSize: 16
  },
  buttonDiv: {
    marginTop: 20
  },
  flexBetweenDiv: {
    display: 'flex',
    justifyContent: 'space-between'
  }
});

/**
 * Showing the order summary information and allow users to scan table number as well as place the order.
 */
export class OrderSummary extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    orders: PropTypes.object.isRequired,
    reduxOrders: PropTypes.object.isRequired,
    clearOrders: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired
  };

  state = {
    isBtnDisable: false,
    isAlertDialogOpen: false,
    isShowProgress: false
  };

  /**
   * Set the isAlertDialogOpen state to the opposite value.
   * @return {null} No return.
   */
  handleToggleAlertDialog = () => this.setState(({ isAlertDialogOpen }) => ({ isAlertDialogOpen: !isAlertDialogOpen }));

  /**
   * Calling the clearOrders action and redirect to the home page
   * @return {null} No Return.
   */
  handleClearOrders = () => {
    this.props.clearOrders();
    this.props.history.push(HOME_PAGE_URL);
  };

  /**
   * Call the placeOrder action and push the page to the order status checking page.
   * @return {null} No return.
   */
  handlePlaceBtnClick = () => {
    this.setState({ isShowProgress: true, isBtnDisable: true });
    return placeOrder(this.props.reduxOrders, this.props.user.jwt).then(data => {
      this.props.clearOrders();
      this.props.history.push(`${ORDER_STATUS_PAGE_URL}/${data}`);
    }).catch(err => console.error(err));
  };

  /**
   * The render method for the component.
   * @return {jsx} Return the jsx.
   */
  render() {
    const { classes, orders, reduxOrders } = this.props;
    const { isAlertDialogOpen, isBtnDisable, isShowProgress } = this.state;
    return (
      <Fragment>
        <Card className={classes.card}>
          <CardContent>
            <Typography className={classes.title} variant="subheading" color="primary">Order summary</Typography>
            <List disablePadding dense>
              {Object.keys(orders.categories).map(category => (
                <ListItem button key={category}>
                  <ListItemText className={classes.categoryTitle} primary={`${category} Ã— ${orders.categories[category].qty}`} />
                  <ListItemSecondaryAction>
                    <Typography color="textSecondary" className={classes.categoryPrice}>${orders.categories[category].price}</Typography>
                  </ListItemSecondaryAction>
                </ListItem>
              ))}
            </List>
            <div className={classes.hrStyle} />
            <div className={classes.flexBetweenDiv}>
              <QRCodeScanner />
              <div>
                <div className={classes.flexEndDiv}><Typography color="primary">Order Price:&nbsp;&nbsp;</Typography><Typography color="textSecondary">${orders.price}</Typography></div>
                <div className={classes.flexEndDiv}><Typography color="primary">Tax:&nbsp;&nbsp;</Typography><Typography color="textSecondary">${orders.tax}</Typography></div>
                <div className={classes.flexEndDiv}><Typography className={classes.fontBold} color="primary">Total Price:&nbsp;&nbsp;</Typography><Typography className={classes.fontBold} color="textSecondary">${orders.totalPrice}</Typography></div>
              </div>
            </div>

            <div className={`${classes.flexEndDiv} ${classes.buttonDiv}`}>
              {isShowProgress && <CircularProgress size={30} thickness={6} />}
              <Button onClick={this.handleToggleAlertDialog} variant="contained" size="small" className={classes.clearBtn}>Clear<RemoveShoppingCart className={classes.icon} /></Button>
              <Tooltip title={reduxOrders.tableNumber ? 'After place your order, the kitchen will start to cook your meal' : 'Please scan the QR code ont the table first'}>
                <div><Button onClick={this.handlePlaceBtnClick} variant="contained" size="small" color="primary" className={classes.placeBtn} disabled={!reduxOrders.tableNumber || isBtnDisable}>Place<Send className={classes.icon} /></Button></div>
              </Tooltip>
            </div>
          </CardContent>
        </Card>
        <AlertDialog
          open={isAlertDialogOpen}
          onClose={this.handleToggleAlertDialog}
          title="Clear all item in the order"
          content="This action will clear all item you has already add in the order."
          onConfirm={this.handleClearOrders}
          confirmButtonText="Clear"
        />
      </Fragment>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  reduxOrders: state.orders,
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  clearOrders: () => dispatch(clearOrders())
});
export default connect(mapStateToProps, mapDispatchToProps)(withRouter((withStyles(styles)(OrderSummary))));
// export default withStyles(styles)(OrderSummary);
