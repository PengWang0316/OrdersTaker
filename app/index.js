import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { AppContainer } from 'react-hot-loader';
import { PersistGate } from 'redux-persist/integration/react';

import App from './components/App';
import configureStore from './store/ConfigureStore';
import './styles/index.global.css';
import registerServiceWorker from './registerServiceWorker';

const { store, persistor } = configureStore();

// ReactDOM.render(
//   <Provider store={store}>
//     <App />
//   </Provider>,
//   document.getElementById('app')
// );
const render = Component => {
  ReactDOM.render(
    <AppContainer>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Component />
        </PersistGate>
      </Provider>
    </AppContainer>,
    document.getElementById('app'),
  );
};
render(App);

// Webpack Hot Module Replacement API
if (module.hot) module.hot.accept('./components/App', () => render(App));

// Registering a service worker in the production enviroment.
registerServiceWorker();
