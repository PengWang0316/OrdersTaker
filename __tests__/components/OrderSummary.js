import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { HOME_PAGE_URL, ORDER_STATUS_PAGE_URL } from '../../app/config';
import { OrderSummary } from '../../app/components/OrderSummary';

jest.mock('@material-ui/core/Card', () => 'Card');
jest.mock('@material-ui/core/CardContent', () => 'CardContent');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/List', () => 'List');
jest.mock('@material-ui/core/CircularProgress', () => 'CircularProgress');
jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/ListItemSecondaryAction', () => 'ListItemSecondaryAction');
jest.mock('@material-ui/core/Tooltip', () => 'Tooltip');
jest.mock('@material-ui/icons/RemoveShoppingCart', () => 'RemoveShoppingCart');
jest.mock('@material-ui/icons/Send', () => 'Send');
jest.mock('../../app/components/QRCodeScanner', () => 'QRCodeScanner');
jest.mock('../../app/components/AlertDialog', () => 'AlertDialog');

jest.mock('../../app/actions/OrdersActions', () => ({ placeOrder: jest.fn().mockReturnValue(Promise.resolve('orderId')) }));

describe('OrderSummary', () => {
  const defaultProps = {
    classes: {
      card: 'card',
      title: 'title',
      categoryTitle: 'categoryTitle',
      flexEndDiv: 'flexEndDiv',
      fontBold: 'fontBold',
      hrStyle: 'hrStyle',
      placeBtn: 'placeBtn',
      ClearBtn: 'clearBtn',
      icon: 'icon',
      buttonDiv: 'buttonDiv',
      FlexBetweenDiv: 'flexBetweenDiv'
    },
    orders: {
      categories: {
        categoryA: { price: 100, tax: 10, qty: 2 }
      }
    },
    reduxOrders: {},
    clearOrders: jest.fn(),
    history: {
      push: jest.fn()
    },
    user: { jwt: 'jwt' }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<OrderSummary {...props} />);

  test('Initial states', () => {
    const { isBtnDisable, isAlertDialogOpen, isShowProgress } = getShallowComponent().state();
    expect(isBtnDisable).toBe(false);
    expect(isAlertDialogOpen).toBe(false);
    expect(isShowProgress).toBe(false);
  });

  test('handleToggleAlertDialog', () => {
    const component = getShallowComponent();
    expect(component.state('isAlertDialogOpen')).toBe(false);
    component.instance().handleToggleAlertDialog();
    expect(component.state('isAlertDialogOpen')).toBe(true);
  });

  test('handleClearOrders', () => {
    getShallowComponent().instance().handleClearOrders();
    expect(defaultProps.clearOrders).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenLastCalledWith(HOME_PAGE_URL);
  });

  test('handlePlaceBtnClick without error', async () => {
    window.console.error = jest.fn();
    const OrdersActions = require('../../app/actions/OrdersActions');
    const component = getShallowComponent();
    await component.instance().handlePlaceBtnClick();
    expect(component.state('isBtnDisable')).toBe(true);
    expect(component.state('isShowProgress')).toBe(true);
    expect(OrdersActions.placeOrder).toHaveBeenCalledTimes(1);
    expect(defaultProps.clearOrders).toHaveBeenCalledTimes(2);
    expect(defaultProps.history.push).toHaveBeenCalledTimes(2);
    expect(defaultProps.history.push).toHaveBeenLastCalledWith(`${ORDER_STATUS_PAGE_URL}/orderId`);
    expect(window.console.error).not.toHaveBeenCalled();
  });

  test('handlePlaceBtnClick with error', async () => {
    window.console.error = jest.fn();
    const OrdersActions = require('../../app/actions/OrdersActions');
    OrdersActions.placeOrder.mockReturnValue(Promise.reject());
    const component = getShallowComponent();
    await component.instance().handlePlaceBtnClick();
    expect(component.state('isBtnDisable')).toBe(true);
    expect(component.state('isShowProgress')).toBe(true);
    expect(OrdersActions.placeOrder).toHaveBeenCalledTimes(2);
    expect(defaultProps.clearOrders).toHaveBeenCalledTimes(2);
    expect(defaultProps.history.push).toHaveBeenCalledTimes(2);
    expect(window.console.error).toHaveBeenCalledTimes(1);
  });

  test('Snapshot without table number', () => expect(renderer.create(<OrderSummary {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with table number', () => expect(renderer.create(<OrderSummary {...{ ...defaultProps, reduxOrders: { tableNumber: 1 } }} />).toJSON()).toMatchSnapshot());
});
