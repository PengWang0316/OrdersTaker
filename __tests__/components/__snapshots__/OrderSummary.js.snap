// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary Snapshot with table number 1`] = `
Array [
  <Card
    className="card"
  >
    <CardContent>
      <Typography
        className="title"
        color="primary"
        variant="subheading"
      >
        Order summary
      </Typography>
      <List
        dense={true}
        disablePadding={true}
      >
        <ListItem
          button={true}
        >
          <ListItemText
            className="categoryTitle"
            primary="categoryA × 2"
          />
          <ListItemSecondaryAction>
            <Typography
              color="textSecondary"
            >
              $
              100
            </Typography>
          </ListItemSecondaryAction>
        </ListItem>
      </List>
      <div
        className="hrStyle"
      />
      <div>
        <QRCodeScanner />
        <div>
          <div
            className="flexEndDiv"
          >
            <Typography
              color="primary"
            >
              Order Price:  
            </Typography>
            <Typography
              color="textSecondary"
            >
              $
            </Typography>
          </div>
          <div
            className="flexEndDiv"
          >
            <Typography
              color="primary"
            >
              Tax:  
            </Typography>
            <Typography
              color="textSecondary"
            >
              $
            </Typography>
          </div>
          <div
            className="flexEndDiv"
          >
            <Typography
              className="fontBold"
              color="primary"
            >
              Total Price:  
            </Typography>
            <Typography
              className="fontBold"
              color="textSecondary"
            >
              $
            </Typography>
          </div>
        </div>
      </div>
      <div
        className="flexEndDiv buttonDiv"
      >
        <Button
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Clear
          <RemoveShoppingCart
            className="icon"
          />
        </Button>
        <Tooltip
          title="After place your order, the kitchen will start to cook your meal"
        >
          <div>
            <Button
              className="placeBtn"
              color="primary"
              disabled={false}
              onClick={[Function]}
              size="small"
              variant="contained"
            >
              Place
              <Send
                className="icon"
              />
            </Button>
          </div>
        </Tooltip>
      </div>
    </CardContent>
  </Card>,
  <AlertDialog
    confirmButtonText="Clear"
    content="This action will clear all item you have already add in the order."
    onClose={[Function]}
    onConfirm={[Function]}
    open={false}
    title="Clear all item in the order"
  />,
]
`;

exports[`OrderSummary Snapshot without table number 1`] = `
Array [
  <Card
    className="card"
  >
    <CardContent>
      <Typography
        className="title"
        color="primary"
        variant="subheading"
      >
        Order summary
      </Typography>
      <List
        dense={true}
        disablePadding={true}
      >
        <ListItem
          button={true}
        >
          <ListItemText
            className="categoryTitle"
            primary="categoryA × 2"
          />
          <ListItemSecondaryAction>
            <Typography
              color="textSecondary"
            >
              $
              100
            </Typography>
          </ListItemSecondaryAction>
        </ListItem>
      </List>
      <div
        className="hrStyle"
      />
      <div>
        <QRCodeScanner />
        <div>
          <div
            className="flexEndDiv"
          >
            <Typography
              color="primary"
            >
              Order Price:  
            </Typography>
            <Typography
              color="textSecondary"
            >
              $
            </Typography>
          </div>
          <div
            className="flexEndDiv"
          >
            <Typography
              color="primary"
            >
              Tax:  
            </Typography>
            <Typography
              color="textSecondary"
            >
              $
            </Typography>
          </div>
          <div
            className="flexEndDiv"
          >
            <Typography
              className="fontBold"
              color="primary"
            >
              Total Price:  
            </Typography>
            <Typography
              className="fontBold"
              color="textSecondary"
            >
              $
            </Typography>
          </div>
        </div>
      </div>
      <div
        className="flexEndDiv buttonDiv"
      >
        <Button
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Clear
          <RemoveShoppingCart
            className="icon"
          />
        </Button>
        <Tooltip
          title="Please scan the QR code ont the table first"
        >
          <div>
            <Button
              className="placeBtn"
              color="primary"
              disabled={true}
              onClick={[Function]}
              size="small"
              variant="contained"
            >
              Place
              <Send
                className="icon"
              />
            </Button>
          </div>
        </Tooltip>
      </div>
    </CardContent>
  </Card>,
  <AlertDialog
    confirmButtonText="Clear"
    content="This action will clear all item you have already add in the order."
    onClose={[Function]}
    onConfirm={[Function]}
    open={false}
    title="Clear all item in the order"
  />,
]
`;
